### ISSUE TITLE
P2588R3 `barrier`'s Phase Completion Guarantees

### ISSUE THREAD

**Issue by StephanTLavavej** [microsoft/STL#3440]
WG21-P2588R3 `barrier`'s Phase Completion Guarantees

**Increased** feature-test macro (expected):
```cpp
#define __cpp_lib_barrier 202302L
```

**Comment by AlexGuteniev** [comment]
I have looked into the wording. This feature relaxes some requirements on Completion Function:
 * There's a freedom to call Completion Function in any thread (including, like, thread pool thread), not only the thread that called last `arrive`
 * There's a freedom not to call Completion Function at all if no thread calls `wait`

I'm not sure if this feature can be usefully utilized in this implementation. 

Apparently this 
> enable `std::barrier` to use hardware accelerators for synchronization, such as those available in NVIDIA GPUs

gives some clue on the targets for this optimization

**Comment by CaseyCarter** [comment]
> I have looked into the wording. This feature relaxes some requirements on Completion Function:
> 
> * There's a freedom to call Completion Function in any thread (including, like, thread pool thread), not only the thread that called last `arrive`

IIRC it can be any thread that called `arrive`, `arrive_and_drop`, or `wait`. (Note that `arrive_and_wait` need not be listed here because it is specified to call both `arrive` and `wait`.)

> * There's a freedom not to call Completion Function at all if no thread calls `wait`
> 
> I'm not sure if this feature can be usefully utilized in this implementation.
> 
> Apparently this
> 
> > enable `std::barrier` to use hardware accelerators for synchronization, such as those available in NVIDIA GPUs
> 
> gives some clue on the targets for this optimization

Yes, the intent of the paper is to allow some implementer freedom without breaking the existing implementations. Programs that call `arrive` but not `wait` become non-portable, but that's weird enough to be considered misuse of `barrier`.

TLDR: we can change our implementation if this relaxation enables us to improve anything, or simply do nothing and declare victory if we like.

**Comment by StephanTLavavej** [comment]
Sounds like we should just declare victory!
